File: src/calLD.cpp
/* 
 * File:   calLD.cpp
 * Author: fan
 * 
 * Created on September 18, 2014, 3:46 PM
 */
#include <iostream>
#include <vector>

#include "calLD.h"
#include "EM.h"
#include "readData.h"

using namespace std;

calLD::calLD() {
}

calLD::calLD(const calLD& orig) {
}

calLD::~calLD() {
}

vector<vector<double> > calLD::calRD(const vector<vector<int> > &genoCode, int windowSize, string ldMeasure, string ldMethod){
    readData rf;
    EM em;

    vector<vector<double> > rSquares; // each element is a window
    
    // cout << "calculate pair-wise r squared...";
    for(int i=0; i < genoCode.size()-windowSize+1; i++){ // total windows
        // cout << "\twindow " << i+1 << "/" << genoCode.size()-windowSize+1 << "...";
        vector<double> t; // contains all LD measures of one window
        vector<double>  tt; // temp vector to reduce calculation to only half matrix
        tt.resize(genoCode.size() * genoCode.size());
        for(int j=i; j< i+windowSize; j++){ // each window
            for(int k=i; k< i+windowSize; k++){ // pair wise
                if(k==j){
                    t.push_back(1.0); // LD with itself
                }else if(k>j){ // upright matrix
                    vector<vector<int> > genoTwoLoci = rf.rmMissingGenotype(genoCode[j],genoCode[k]);
                    double rSquare;
                    if(ldMeasure=="r2"){
                        if(ldMethod=="rh"){
                            rSquare = em.rh_Rsquared(genoTwoLoci[0],genoTwoLoci[1]);
                        }else if(ldMethod=="esem"){
                            rSquare = em.esem_Rsquared(genoTwoLoci[0],genoTwoLoci[1]);
                        }                       
                    }else{
                        if(ldMethod=="rh"){
                            rSquare = em.rh_Dprime(genoTwoLoci[0],genoTwoLoci[1]);
                        }else if(ldMethod=="esem"){
                            rSquare = em.esem_Dprime(genoTwoLoci[0],genoTwoLoci[1]);
                        }                         
                    }                    
                    t.push_back(rSquare);  
                    tt[genoCode.size() * j + k] =rSquare;
                }else{ // bottom left matrix
                    t.push_back(tt[genoCode.size() * k + j]);
                }
            }
        }
        rSquares.push_back(t);
        // cout << "done" << endl;
    }
    // cout << "calculate pair-wise r squared ... done" << endl;
    // cout << "done" << endl;
    return rSquares;
}

vector<vector<double> > calLD::calRD2(const vector<vector<int> > &genoCode, int windowSize, string ldMeasure, string ldMethod){
    readData rf;
    EM em;

    vector<vector<double> > rSquares; // each element is a window
    
    // temp vector to reduce calculation to only half matrix
    // matrix for window 0 or previous window for second to last windows
    vector<double>  tt; 
    tt.resize(windowSize * windowSize);
        
    // cout << "calculate pair-wise r squared...";
    for(int i=0; i < genoCode.size()-windowSize+1; i++){ // total windows
        vector<double> t; // contains all LD measures of one window 
        
        // temp vector to reduce calculation to only half matrix            
        vector<double>  tt0; 
        tt0.resize(windowSize * windowSize);
        
        // each window
        if(i==0){ // first window
            for(int j=i; j< i+windowSize; j++){ // each row
                for(int k=i; k< i+windowSize; k++){ // each column
                    if(k==j){ // diagonal
                        t.push_back(1.0); // LD with itself
                        tt[windowSize * (j-i) + k-i] = 1.0;
                    }else if(k>j){ // upright matrix
                        vector<vector<int> > genoTwoLoci = rf.rmMissingGenotype(genoCode[j],genoCode[k]);
                        double rSquare;
                        if(ldMeasure=="r2"){
                            if(ldMethod=="rh"){
                                rSquare = em.rh_Rsquared(genoTwoLoci[0],genoTwoLoci[1]);
                            }else if(ldMethod=="esem"){
                                rSquare = em.esem_Rsquared(genoTwoLoci[0],genoTwoLoci[1]);
                            }                       
                        }else{
                            if(ldMethod=="rh"){
                                rSquare = em.rh_Dprime(genoTwoLoci[0],genoTwoLoci[1]);
                            }else if(ldMethod=="esem"){
                                rSquare = em.esem_Dprime(genoTwoLoci[0],genoTwoLoci[1]);
                            }                         
                        }                    
                        t.push_back(rSquare);  
                        tt[windowSize * (j-i) + k-i] = rSquare;
                    }else{ // bottom left matrix
                        tt[windowSize * (j-i) + k-i] = tt[windowSize * (k-i) + j-i];
                        t.push_back(tt[windowSize * (j-i) + k-i]);
                    }
                }
            }
        }else{ // second to last window           
            for(int j=i; j< i+windowSize; j++){ // each row
                for(int k=i; k< i+windowSize; k++){ // each column
                    if(k==j){ // diagonal
                        t.push_back(1.0); // LD with itself
                        tt0[windowSize * (j-i) + k-i] = 1.0;
                    }else if(k>j){ // upright matrix
                        if(k == i + windowSize -1){ // only calculate last column
                            vector<vector<int> > genoTwoLoci = rf.rmMissingGenotype(genoCode[j],genoCode[k]);
                            double rSquare;
                            if(ldMeasure=="r2"){
                                if(ldMethod=="rh"){
                                    rSquare = em.rh_Rsquared(genoTwoLoci[0],genoTwoLoci[1]);
                                }else if(ldMethod=="esem"){
                                    rSquare = em.esem_Rsquared(genoTwoLoci[0],genoTwoLoci[1]);
                                }                       
                            }else{
                                if(ldMethod=="rh"){
                                    rSquare = em.rh_Dprime(genoTwoLoci[0],genoTwoLoci[1]);
                                }else if(ldMethod=="esem"){
                                    rSquare = em.esem_Dprime(genoTwoLoci[0],genoTwoLoci[1]);
                                }                         
                            }
                            t.push_back(rSquare);
                            tt0[windowSize * (j-i) + k-i] = rSquare;
                        }else{ // copy from previous matrix for other columns                            
                            tt0[windowSize * (j-i) + k-i] = tt[windowSize * (j-i+1) + k-i+1]; // tt0[i][j] = tt[i+1][j+1]
                            t.push_back(tt0[windowSize * (j-i) + k-i]);
                        }                        
                    }else{ // bottom left matrix
                        tt0[windowSize * (j-i) + k-i] = tt0[windowSize * (k-i) + j-i];
                        t.push_back(tt0[windowSize * (j-i) + k-i]);
                    }
                }
            }
            tt.clear();
            tt = tt0;
            tt0.clear();
        }        
        
        rSquares.push_back(t);
        t.clear();
    }

    return rSquares;
}

File: src/CalLD.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package quantld;


/**
 *
 * @author Felix Yanhui Fan felixfanyh@gmail.com
 */
public class CalLD {
    private double tol;
    private int maxItr;
    
    // default value
    {
        tol = 0.001;
        maxItr = 1000;
    }
    
    public CalLD(){
    }
    
    public CalLD(double tol,int maxItr){
        this.tol = tol;
        this.maxItr = maxItr;
    }
    
    private final EM em = new EM(tol, maxItr);
    
    public double[][][] CalRDprime(int[][] genoCode, int windowSize, String ldMeasure){   
        
        // total number of windows
        int totWindows = genoCode.length - windowSize + 1;
        
        // output, r2 or D prime, each window has a r2 matrix
        double[][][] rSquares = new double[totWindows][windowSize][windowSize];
        
        // calculate r2 or D prime for each window
        // first window, calculate all matrix 
        for(int i=0; i<windowSize;i++){ // first snp
            for(int j=0; j<windowSize;j++){ // second snp
                if(i == j){ // diagonal, LD with itself
                    rSquares[0][i][j] = 1.0;
                }else if(i<j){ // upright matrix
                    switch (ldMeasure) {
                        case "r2":
                            rSquares[0][i][j] = em.esemR2(genoCode[i],genoCode[j]);
                            break;
                        case "dp":
                            rSquares[0][i][j] = em.esemDprime(genoCode[i],genoCode[j]);
                            break;
                        case "sr2":
                            rSquares[0][i][j] = em.esemSignedR2(genoCode[i],genoCode[j]);
                            break;
                        default:
                            System.out.println("--ld-measure only support r2, dp or sr2");
                            System.exit(0);
                    } 
                }else{ // bottom left matrix 
                    rSquares[0][i][j] = rSquares[0][j][i];
                }
            }
        }
        // second to last windows, only calculate the last row and last column of the matrix
        // other values were already calculated in previous window
        for(int n=1; n<totWindows; n++){
            for(int i=0; i<windowSize;i++){ // first snp
                for(int j=0; j<windowSize;j++){ // second snp
                    if(i < windowSize-1 && j < windowSize-1){// were already calculated in previous matrix 
                        rSquares[n][i][j] = rSquares[n-1][i+1][j+1];
                    }else{
                        if(i<j){// upright matrix, last column
                            switch (ldMeasure) {
                                case "r2":
                                    rSquares[n][i][j] = em.esemR2(genoCode[n+i],genoCode[n+j]);
                                    break;
                                case "dp":
                                    rSquares[n][i][j] = em.esemDprime(genoCode[n+i],genoCode[n+j]);
                                    break;
                                case "sr2":
                                    rSquares[n][i][j] = em.esemSignedR2(genoCode[n+i],genoCode[n+j]);
                                    break;
                                default:
                                    System.out.println("--ld-measure only support r2, dp or sr2");
                                    System.exit(0);
                            }
                        }else if(i==j){
                            rSquares[n][i][j] = 1;
                        }else{// bottom left matrix, last row
                            rSquares[n][i][j] = rSquares[n][j][i];
                        }
                    }
                }
            }
            
        }
        return rSquares;
    }
}
